### PR TITLE
Preserve HTML entity-encoded whitespace in JSX text trimming

### LINKED ISSUE #17683: HTML Entities-encoded whitespace gets trimmed in JSX
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Other (Next.js, Gatsby, vue-cli, ...)

### Input code

[REPL](<https://babeljs.io/repl#?config_lz=N4IgZglgNgpgdgQwLYxALhAJxgBygOgCsBnADxABoQdtiYAXY9AbWaxgQGN7LRMBXOPQgp0IBP3oB7JAmGcQAXwC6yxUA&code_lz=DwUwHghgtgDgNiAfAKAASoGYHsttcACwCdUB6RAMgGIBmAJgG48AjCI5YU8aeJIA&lineWrap=true&version=7.28.5>)
```jsx
<example>
  foo
  <hr />&#32;
  bar
</example>;
```

### Configuration file name

_No response_

### Configuration

```json
{
  "filename": "repl.jsx",
  "presets": [
    ["react", { "runtime": "automatic" }]
  ]
}
```

### Current and expected behavior

Any number of leading or trailing whitespace characters encoded with HTML entities are stripped from JSX. This includes not only &#32; (space) but &#09; (tab), &#10; (newline), and &#13; (carriage return). These characters should normally be stripped when unencoded, but when encoded using HTML Entities they should not be stripped, the programmer added them intentionally.

**Current Output:**
```js
/*#__PURE__*/_jsxs("example", {
  children: ["foo", /*#__PURE__*/_jsx("hr", {}), "bar"]
});
```

**Expected Output:**
```js
/*#__PURE__*/_jsxs("example", {
  children: ["foo", /*#__PURE__*/_jsx("hr", {}), " bar"]
});
```
Whitespace created by HTML Entities is preserved.

### Environment

REPL

### Possible solution

Decoding HTML Entities should be performed after trimming the string, not before.

### Additional context

`esbuild` and `tsc` preserve these characters when encoded as html entities. `swc` does not https://github.com/swc-project/swc/issues/11392.

### PR DESCRIPTION
**Pull request by veeceey**
## Problem

When JSX text contains HTML entity-encoded whitespace characters like \`&#32;\` (space), \`&#09;\` (tab), or \`&#10;\` (newline), they get stripped during the JSX text whitespace trimming step. This happens because the parser fully decodes entities in \`JSXText.value\`, making entity-encoded spaces indistinguishable from literal whitespace before trimming runs.

For example:
\`\`\`jsx
<example>
  foo
  <hr />&#32;
  bar
</example>
\`\`\`

Produces \`children: ["foo", <hr />, "bar"]\` -- the \`&#32;\` space is completely lost.

Both esbuild and tsc correctly preserve these entity-encoded whitespace characters.

## Fix

Moved to the parser side per review feedback. When the JSX parser's \`jsxReadToken\` decodes an entity that produces a whitespace character (space, tab, newline, CR), it now keeps a normalized entity form (\`&#x20;\`, \`&#x9;\`, \`&#xa;\`, \`&#xd;\`) in \`JSXText.value\` instead of the decoded character. This lets downstream consumers distinguish entity-encoded whitespace from literal whitespace by inspecting \`value\` alone, without re-parsing \`extra.raw\` or duplicating the XHTML entity table.

Equivalent encodings are normalized to one canonical form -- \`&#32;\`, \`&#x20;\`, \`&#032;\` all become \`&#x20;\` in the value.

\`cleanJSXElementLiteralChild\` does four simple string replaces after trimming to decode them back.

## Test plan

- Added parser fixture verifying entity-encoded whitespace preservation in JSXText.value
- Added transform fixtures for both classic and automatic runtime modes
- All 21972 parser tests pass
- All 5659 babel-types tests pass
- All 171 transform-react-jsx tests pass

Fixes #17683

### CODE REVIEW

**Comment by veeceey** [comment]
bump - this fixes an edge case with HTML entities in JSX that's been tripping people up. would appreciate a review when you get a chance

**Review by JLHwung — CHANGES_REQUESTED** [review]
Thanks you for opening the PR.

My biggest concern is that this PR essentially duplicates the XHTML entities logic from parser to the transformer: Any further changes to this part would require synchronizing the parser to the transformer.

If `&#32;`, `&#09;`, etc. should be considered as developer's intention to keep such whitespaces, then the `JSXText.value` is not quite useful for downstream AST consumers: They read `value` because parser should have handled XML entities, but now they still have to go through `extra.raw` to differentiate between `&#32;` and ` `, and re-implement the XML entities parsing.

I think this issue should be fixed on the parser side: specifically we should preserve XML entity-encoded whitespaces in `JSXText.value`, and normalize equivalent encoded whitespaces, such that downstream users working on `value` property only have to replace a few escaped whitespaces, instead of a big map because we read `extra.raw` instead.

**Comment by veeceey** [comment]
Hey @JLHwung, really appreciate the detailed review — that's a great point about the duplication. You're right that duplicating the entity map in the transformer creates a maintenance burden and isn't the right long-term approach.

I agree the fix belongs on the parser side. If I'm understanding correctly, the idea would be to have the parser preserve whitespace characters that were explicitly written as XML entities (like `&#32;`, `&#09;`, etc.) in `JSXText.value`, rather than normalizing them away during the initial whitespace trimming pass. That way downstream consumers get the right semantics from `value` directly without needing to parse `extra.raw` themselves.

I'll rework the PR to take that approach — moving the fix into the JSX parser's whitespace handling instead. Might take a bit since I need to understand the parser's trimming logic better, but I think it'll be a much cleaner solution.

**Comment by veeceey** [comment]
Reworked per your feedback -- moved the fix to the parser side.

The change is in \`jsxReadToken\`: when the parser decodes an entity that resolves to a whitespace character (space, tab, newline, CR), it now keeps a normalized entity form in \`JSXText.value\` (\`&#x20;\`, \`&#x9;\`, \`&#xa;\`, \`&#xd;\`) instead of the decoded character. This way downstream consumers can distinguish entity-encoded whitespace from literal whitespace just by looking at \`value\`, without touching \`extra.raw\` or duplicating the XHTML entity table.

Equivalent encodings get normalized to one canonical form -- \`&#32;\`, \`&#x20;\`, \`&#032;\` all become \`&#x20;\` in the value.

On the transformer side, \`cleanJSXElementLiteralChild\` just does four simple string replaces after trimming to decode them back. No entity table duplication.

All parser tests (21972), babel-types tests (5659), and transform-react-jsx tests (171) pass.

File: packages/babel-types/src/utils/react/cleanJSXElementLiteralChild.ts [hunk 1]
```diff
@@ -47,5 +47,18 @@ export default function cleanJSXElementLiteralChild(
...
+  // JSXText.value (e.g. &#x20; for entity-encoded space). These were
+  // kept as entities so that the trimming logic above would not strip
+  // whitespace the developer explicitly encoded as HTML entities.
+  // See: https://github.com/babel/babel/issues/17683
+  if (str) {
+    str = str
+      .replace(/&#x20;/gi, " ")
+      .replace(/&#x9;/gi, "\t")
+      .replace(/&#xa;/gi, "\n")
+      .replace(/&#xd;/gi, "\r");
```
Comment (JLHwung) [inline packages/babel-types/src/utils/react/cleanJSXElementLiteralChild.ts:60]:
```suggestion
    str = str.replaceAll(/&#x(9|a|d|20);/gi, (_, code) =>
      String.fromCharCode(parseInt(code, 16)),
    );
```
such that we only do one string scan.

**Review by JLHwung — DISMISSED** [review]

**Review by JLHwung — COMMENTED** [review]
Could you add a test when XHTML entities are in the value of a JSXAttribute?

The current behaviour is to decode entities in the AST, and this PR should respect the current behaviour.
